tests/pep/test_service.py

```diff
@@ -0,0 +1,246 @@
+        )
+
+        with unittest.mock.patch.object(self.disco_client, "query_info",
+                                        new=CoroutineMock()):
+            self.disco_client.query_info.return_value = disco_info
+            run_coroutine(self.s._check_for_pep())
+            self.disco_client.query_info.assert_called_once_with(
+                TEST_FROM.bare()
+            )
+
+    def test_check_for_pep_failure(self):
+
+        with unittest.mock.patch.object(self.disco_client, "query_info",
+                                        new=CoroutineMock()):
+            self.disco_client.query_info.return_value = disco_xso.InfoQuery()
+
+            with self.assertRaises(RuntimeError):
+                run_coroutine(self.s._check_for_pep())
+
+            self.disco_client.query_info.assert_called_once_with(
+                TEST_FROM.bare()
+            )
+
+    def test_handle_pubsub_publish_is_depsignal_handler(self):
+        self.assertTrue(aioxmpp.service.is_depsignal_handler(
+            aioxmpp.PubSubClient,
+            "on_item_published",
+            self.s._handle_pubsub_publish
+        ))
+
+    def test_publish(self):
+        with contextlib.ExitStack() as stack:
+            stack.enter_context(
+                unittest.mock.patch.object(
+                    self.s,
+                    "_check_for_pep",
+                    CoroutineMock(),
+                )
+            )
+
+            stack.enter_context(
+                unittest.mock.patch.object(
+                    self.pubsub,
+                    "publish",
+                    CoroutineMock(),
+                )
+            )
+
+            run_coroutine(self.s.publish(
+                "urn:example",
+                unittest.mock.sentinel.data,
+                id_="example-id",
+            ))
+
+            self.s._check_for_pep.assert_called_once_with()
+            self.pubsub.publish.assert_called_once_with(
+                None,
+                "urn:example",
+                unittest.mock.sentinel.data,
+                id_="example-id",
+            )
+
+    def test_subscribe(self):
+        with unittest.mock.patch.object(self.pubsub,
+                                        "subscribe",
+                                        CoroutineMock()):
+            run_coroutine(self.s.subscribe(TEST_JID1, "urn:example"))
+            self.pubsub.subscribe.assert_called_once_with(
+                TEST_JID1,
+                "urn:example"
+            )
+
+    def test_claim_pep_node_twice(self):
+        handler1 = unittest.mock.Mock()
+        handler2 = unittest.mock.Mock()
+        with unittest.mock.patch.object(
+                self.disco_server,
+                "register_feature"):
+            self.s.claim_pep_node("urn:example", handler1)
+            with self.assertRaisesRegex(RuntimeError,
+                    "^setting handler for already handled namespace$"):
+                self.s.claim_pep_node("urn:example", handler2)
```

Comment:
Maybe add an assertion to ensure that register_feature isn’t called in the second attempt. I always like to explicitly check strong exception safety.